{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.chdir(os.path.expanduser(\"~/speos/\"))\n",
    "from speos.preprocessing.datasets import DatasetBootstrapper\n",
    "from speos.utils.config import Config\n",
    "\n",
    "config = Config()\n",
    "\n",
    "path_to_config = \"config_uc_only_nohetio_film_newstorage.yaml\"\n",
    "# path_to_config = \"config_cad_only_film_newstorage.yaml\"  # change this to your config path\n",
    "config.parse_yaml(path_to_config)\n",
    "\n",
    "dataset = DatasetBootstrapper(name=config.name, config=config).get_dataset()\n",
    "dataset.preprocessor.build_graph(adjacency=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset.preprocessor.hgnc2id[\"MAML2\"]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load pre-computed explanations\n",
    "\n",
    "You will need to change the gene you want to inspect and the path to the explanation .pt files generated by the explanation script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "\n",
    "gene = \"HNF4A\"\n",
    "num_outer = 11\n",
    "num_inner = 10\n",
    "ig_attr_self_all = None\n",
    "ig_attr_self_abs_all = None\n",
    "num_processed = 0\n",
    "for outer_fold in range(0, num_outer):\n",
    "    for inner_fold in range(0, num_inner):\n",
    "        if inner_fold == outer_fold:\n",
    "            continue\n",
    "        try:\n",
    "            #print(\"looking for /mnt/storage/speos/results/{}_ig_attr_self_outer{}_inner{}_{}.pt\".format(config.name, outer_fold, inner_fold, gene))\n",
    "            ig_attr_self = torch.load(\"/mnt/storage/speos/explanations/{}_ig_attr_self_outer{}_inner{}_{}.pt\".format(config.name, outer_fold, inner_fold, gene)).detach().cpu()\n",
    "            #ig_attr_self_abs = torch.load(\"/mnt/storage/speos/results/{}_ig_attr_self_abs_outer{}_inner{}_{}.pt\".format(config.name,outer_fold, inner_fold, gene))\n",
    "            print(\"Loaded attributes for outer {} inner {}\".format(outer_fold, inner_fold))\n",
    "            ig_attr_self.requires_grad = False\n",
    "            #ig_attr_self_abs.requires_grad = False\n",
    "        except FileNotFoundError:\n",
    "            print(\"Attributes for outer {} inner {} not yet calculated.\".format(outer_fold, inner_fold))\n",
    "            continue \n",
    "\n",
    "        num_processed += 1\n",
    "\n",
    "        #if ig_attr_self_abs_all is None:\n",
    "        #    ig_attr_self_abs_all = ig_attr_self_abs\n",
    "        #else:\n",
    "        #    ig_attr_self_abs_all += ig_attr_self_abs\n",
    "\n",
    "        if ig_attr_self_all is None:\n",
    "            ig_attr_self_all = ig_attr_self\n",
    "        else:\n",
    "            ig_attr_self_all += ig_attr_self\n",
    "\n",
    "ig_attr_self_all /= num_processed\n",
    "#ig_attr_self_abs_all /= num_processed\n",
    "\n",
    "input_attributions = ig_attr_self_all.cpu().numpy().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Or easier in one step\n",
    "import torch\n",
    "\n",
    "gene = \"HNF4A\"\n",
    "\n",
    "input_attributions = torch.load(\"/mnt/storage/speos/explanations/{}_ig_attr_self_total_{}.pt\".format(config.name, gene)).detach().cpu().numpy().tolist()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Now prepare the input data for visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "features = dataset.preprocessor.get_feature_names()\n",
    "\n",
    "normalized_data = torch.empty(dataset.data.x.shape, dtype=torch.double)\n",
    "normalized_data_positives = dataset.data.x / dataset.data.x.topk(dim=0, k =int(dataset.data.x.shape[0]/100), sorted=True).values[-1,:]\n",
    "normalized_data_negatives = (dataset.data.x / dataset.data.x.topk(dim=0, k=int(dataset.data.x.shape[0]/100), sorted=True, largest=False).values[-1,:]) * -1\n",
    "\n",
    "normalized_data[dataset.data.x >= 0] = normalized_data_positives[dataset.data.x >= 0]\n",
    "normalized_data[dataset.data.x < 0] = normalized_data_negatives[dataset.data.x < 0]\n",
    "\n",
    "q001 = normalized_data.quantile(0.01, dim=0)\n",
    "q01 = normalized_data.quantile(0.1, dim=0)\n",
    "q25 = normalized_data.quantile(0.25, dim=0)\n",
    "q50 = normalized_data.quantile(0.5, dim=0)\n",
    "q75 = normalized_data.quantile(0.75, dim=0)\n",
    "q90 = normalized_data.quantile(0.9, dim=0)\n",
    "q99 = normalized_data.quantile(0.99, dim=0)\n",
    "\n",
    "input_values = normalized_data[dataset.preprocessor.hgnc2id[gene]].tolist()\n",
    "#input_values = normalized_data[int(gene)].tolist()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sort everything by the highest explainer attribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list1, list2, list3,  q001_ordered, q01_ordered, q25_ordered, q50_ordered, q75_ordered, q90_ordered, q99_ordered= zip(*sorted(zip(input_attributions, features, input_values, q001, q01, q25, q50, q75, q90, q99)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib as plt\n",
    "\"\"\"\n",
    "import matplotlib.font_manager as font_manager\n",
    "\n",
    "# uncomment this part if you also want to set the font to Helvetica. However, you must first download it and place it in your font directory, which can be tricky\n",
    "\n",
    "from matplotlib import get_cachedir\n",
    "from glob import glob\n",
    "\n",
    "font_dirs = ['/mnt/storage/anaconda3/envs/speos/lib/python3.10/site-packages/matplotlib/mpl-data/matplotlibrc']\n",
    "font_files = font_manager.findSystemFonts(fontpaths=font_dirs)\n",
    "\n",
    "for font_file in font_files:\n",
    "    font_manager.fontManager.addfont(font_file)\n",
    "\n",
    "dir_cache = get_cachedir()\n",
    "for file in glob(f'{dir_cache}/*.cache') + glob(f'{dir_cache}/font*'):\n",
    "    if not os.path.isdir(file): # don't dump the tex.cache folder... because dunno why\n",
    "        os.remove(file)\n",
    "        print(f'Deleted font cache {file}.')\n",
    "\n",
    "# set font\n",
    "plt.rcParams['font.family'] = 'Helvetica'\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "full_width = 18\n",
    "cen = 1/2.54\n",
    "small_font = 6\n",
    "medium_font = 8\n",
    "large_font = 10\n",
    "plt.rc('xtick', labelsize=small_font)\n",
    "plt.rc('ytick', labelsize=small_font)\n",
    "plt.rcParams['axes.linewidth'] = 0.4\n",
    "plt.rcParams['ytick.major.size'] = 3\n",
    "plt.rcParams['ytick.major.width'] = 0.5\n",
    "plt.rcParams['ytick.minor.size'] = 2\n",
    "plt.rcParams['ytick.minor.width'] = 0.3\n",
    "plt.rcParams['xtick.major.size'] = 2\n",
    "plt.rcParams['xtick.major.width'] = 0.3\n",
    "plt.rcParams['xtick.minor.size'] = 1\n",
    "plt.rcParams['xtick.minor.width'] = 0.1\n",
    "\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# And visualize it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import numpy as np\n",
    "from matplotlib import cm\n",
    "from numpy import linspace\n",
    "import os\n",
    "\n",
    "orange = '#fd5800'\n",
    "gray = \"#71797E\"\n",
    "\n",
    "get_top_n = 10\n",
    "get_last_n = 5\n",
    "\n",
    "labels = list2[:get_last_n] + list2[-get_top_n:]\n",
    "attributions = list1[:get_last_n] + list1[-get_top_n:]\n",
    "values = list3[:get_last_n] + list3[-get_top_n:]\n",
    "\n",
    "start = 0.0\n",
    "stop = 1.0\n",
    "\n",
    "number_of_lines = 7\n",
    "alpha = 0.5\n",
    "cm_subsection = linspace(stop, start, number_of_lines)\n",
    "colors = [cm.viridis(x) for x in cm_subsection]\n",
    "colors = [color[:3] + (alpha,) for color in colors]\n",
    "\n",
    "q001_, q01_, q25_, q50_, q75_, q90_, q99_ =  [liste[:get_last_n] + liste[-get_top_n:] for liste in [q001_ordered, q01_ordered, q25_ordered, q50_ordered, q75_ordered, q90_ordered, q99_ordered]]\n",
    "\n",
    "labels = [\" \".join([word[0].upper() + word[1:] for word in label.split(\" \")]) for label in labels]\n",
    "labels = [\"-\".join([word[0].upper() + word[1:] for word in label.split(\"-\")]) for label in labels]\n",
    "labels = [\"Z-Score \" + label[6:]if label.startswith(\"ZSTAT\") else label for label in labels]\n",
    "labels = [\"P-Value \" + label[2:] if label.startswith(\"P \") else label for label in labels]\n",
    "labels = [\"n SNPs \" + label[6:] if label.startswith(\"NSNPS\") else label for label in labels]\n",
    "labels = [label + \"\\n(GWAS)\" if label.startswith((\"Z-Score\", \"P-Value\", \"n SNPs\")) else label + \"\\n(Gene Expr.)\" for label in labels]\n",
    "labels = [label[8:] if label.startswith(\"Cells\") else label for label in labels]\n",
    "\n",
    "attributions /= np.max(np.abs(attributions))\n",
    "\n",
    "x = np.arange(len(labels))  # the label locations\n",
    "\n",
    "width = 0.35  # the width of the bars\n",
    "\n",
    "gs_kw = dict(height_ratios=[1, 20])\n",
    "\n",
    "fig, (ax2, ax) = plt.subplots(2,1, figsize=((full_width/2.5)*cen, 11*cen), gridspec_kw=gs_kw)\n",
    "ax.grid(axis=\"y\", color=\"lightgray\", linewidth=0.5)\n",
    "rects2 = ax.barh(x + width/2, attributions, width, color=orange, label='Importance')\n",
    "rects1 = ax.barh(x - width/2, values, width, color=gray, label='Input Value')\n",
    "\n",
    "\n",
    "# Add some text for labels, title and custom x-axis tick labels, etc.\n",
    "#ax.set_ylabel('Scores')\n",
    "#ax.set_title('Scores by group and gender')\n",
    "ax.set_yticks(x, labels)\n",
    "#ax.legend(fontsize=small_font)\n",
    "ax.axvline(x=1, ymin=0, ymax=len(labels)-2, linestyle=\":\", color =\"lightgray\", zorder = -1)\n",
    "ax.axvline(x=0, ymin=0, ymax=len(labels)-2, color =\"gray\", zorder = 0)\n",
    "#ax.bar_label(rects1, padding=3)\n",
    "#ax.bar_label(rects2, padding=3)\n",
    "steps = []\n",
    "steps.append(ax.step(x=q001_, y =x + 0.5, where = \"post\", c=\"lightgray\",zorder=-1, linewidth=0.5)[0])\n",
    "steps.append(ax.step(x=q01_, y =x + 0.5, where = \"post\", c=\"lightgray\", zorder=-1, linewidth=0.5)[0])\n",
    "steps.append(ax.step(x=q25_, y =x +0.5, where = \"post\", c=\"lightgray\", zorder=-1, linewidth=0.5)[0])\n",
    "steps.append(ax.step(x=q50_, y =x +0.5, where = \"post\", c=\"lightgray\", zorder=-1, linewidth=0.5)[0])\n",
    "steps.append(ax.step(x=q75_, y =x +0.5, where = \"post\", c=\"lightgray\", zorder=-1, linewidth=0.5)[0])\n",
    "steps.append(ax.step(x=q90_, y =x +0.5, where = \"post\", c=\"lightgray\", zorder=-1, linewidth=0.5)[0])\n",
    "steps.append(ax.step(x=q99_, y =x +0.5, where = \"post\", c=\"lightgray\", zorder=-1, linewidth=0.5)[0])\n",
    "\n",
    "for i in range(1, len(steps)):\n",
    "    ax.fill_betweenx(y=[value-0.5 if index %2 == 0 else value+0.5 for index, value in enumerate(np.repeat(x,2))], \n",
    "                     x1=np.repeat(steps[i-1].get_xdata(), 2), \n",
    "                     x2=np.repeat(steps[i].get_xdata(), 2), zorder=-1,\n",
    "                     color =colors[i-1],\n",
    "                     linewidth=0.5)\n",
    "\n",
    "ax.set_ylim((-0.5, len(labels) -0.5))         \n",
    "ax.set_xlabel(\"Normalized Input Value and Importance\", size=medium_font)\n",
    "cmap = (mpl.colors.ListedColormap(colors[:len(steps)-1]).with_extremes(over='white', under='white'))\n",
    "\n",
    "\n",
    "bounds = [0.01, 0.1, 0.25, 0.5, 0.75, 0.9, 0.99]\n",
    "norm = mpl.colors.BoundaryNorm(bounds, cmap.N)\n",
    "cbar = fig.colorbar(\n",
    "    mpl.cm.ScalarMappable(cmap=cmap, norm=norm),\n",
    "    cax=ax2,\n",
    "    extend='both',\n",
    "    ticks=bounds,\n",
    "    spacing='proportional',\n",
    "    orientation='horizontal',\n",
    "    label=\"Normalized Quantiles across all Genes' Input Values\"\n",
    ")\n",
    "\n",
    "cbar.ax.set_xticklabels([\".01\", \".1\", \".25\", \".5\", \".75\", \".9\",\".99\"], size=small_font) \n",
    "cbar.ax.set_xlabel(\"Normalized Quantiles across all Genes' Input Values\", size=small_font) \n",
    "\n",
    "fig.tight_layout()\n",
    "plt.savefig(\"Input_explanation_{}.svg\".format(gene), bbox_inches=\"tight\", dpi=300)\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
